text/0000-contextual-helpers.md

```diff
@@ -0,0 +1,122 @@
+of this in the wild can be found in [ember-basic-dropdown](https://github.com/cibernox/ember-basic-dropdown),
+[ember-form-for](https://github.com/martndemus/ember-form-for) and [ember-power-calendar](https://github.com/cibernox/ember-power-calendar).
+
+However, the same feature is not yet supported for helpers and it would be equally useful for
+exposing helpers with prepopulated arguments or options to provide better APIs.
+
+# Detailed design
+
+The design of this feature should be as similar to the **contextual components** feature as
+possible so users feel confortable with it immediatly.
+
+Therefore the natural option would be the creation of a `helper` helper that depending on the context
+invokes a helper by its name or creates a closure helper to pass it around.
+
+However there is an ambiguity with this approach that is not present with components. I'm going to expose
+the problem first since it's going to drive the entire design of the feature.
+
+```hbs
+{{my-component format-date=(helper "moment-format")}}
+```
+
+In the example above it's not clear the intention of the developer. The developer's intention
+can be a few things:
+
+- Immediatly evaluate the helper and pass the value of its evaludation to the component
+- Create a closure helper and pass it to the component, so it can be called later.
+
+Note that this ambiguity is not present when the `helper` helper is called in the
+context of html. By example
+
+```hbs
+<strong>My name is {{helper 'capitalize-str' user.name}}</strong>
+<button title={{helper 'capitalize-str' user.name}}></button>
+```
+
+In those examples above the only option can be immediatly evaluate.
+
+In the proposed syntax of glimmer-components the ambiguity doesn't exists when passing attributes
+but it does when passing properties:
+
+```hbs
+<my-component class={{helper 'dasherize-str' className}}></my-component> <!-- Immediate evaluation -->
+<my-component @class={{helper 'dasherize-str' className}}></my-component> <!-- Unclear. Can be evaluation or closure components-->
+```
+
+How is this problem solved?
+
+#### Use a different helper for creating closure helpers and to evaluate helpers
+
+To remove the ambiguity, there is two different keywords for the two different features.
+
+Ideas include `helper` to create the closure helper and `invoke-helper` to invoke it.
+Another possible names borrowed from ruby conventions can be `helper` (closure creator) and `helper!` (invocation).
+
+Examples:
+
+```hbs
+<strong>My name is {{invoke-helper 'capitalize-str' user.name}}</strong>
+<button title={{invoke-helper 'capitalize-str' user.name}}></button>
+<strong>My name is {{helper 'capitalize-str' user.name}}</strong> <!-- Invalid invocation -->
+<button title={{helper 'capitalize-str' user.name}}></button> <!-- Invalid invocation -->
+
+{{my-component format-date=(helper "moment-format")}} <!-- Creates closure helper -->
+{{my-component format-date=(invoke-helper "moment-format")}} <!-- Invokes helper -->
```

Comment:
that would invoke the helper which generates a string that is passed to the component, like today.